087-external-certificate-manager.md

```diff
@@ -0,0 +1,241 @@
+  The `name`, `kind`, and `group` properties will be copied over into the `Certificate` custom resource Strimzi creates.
+3. The property `publicCert` will only be used by Strimzi if `type` is set to `cert-manager.io`.
+  The `secretName` and `certificate` properties will be used to locate the CA public certificate that must be trusted by Strimzi components in order to trust the end-entity certificates that cert-manager issues.
+
+### User steps
+
+To make use of this new option the user will have to:
+
+1. Install cert-manager.
+2. Create an `Issuer` or `ClusterIssuer` custom resource.
+3. Create a `Secret` containing the CA public cert for Strimzi to trust.
+   Users can optionally use [trust-manager](https://cert-manager.io/docs/trust/trust-manager/) to create this Secret, but they are responsible for installing trust-manager and creating the `Bundle` CR with a single `sources` entry.
+4. Create a `Kafka` resource with `clusterCa.type` and/or `clientsCa.type` set to `cert-manager.io`, and `certManagerIssuerRef` and `publicCert` configured.
+
+### Internal handling of certificates
+
+Strimzi uses two generation annotations to track how Cluster CA certificates are trusted and used:
+* `strimzi.io/cluster-ca-key-generation` on Kafka pods to indicate the CA generation trusted by that pod
+* `strimzi.io/cluster-ca-cert-generation` on Secrets containing certs to indicate the CA generation that signed those certs
+
+The `strimzi.io/cluster-ca-cert-generation` annotation is also used on Kafka pods to indicate the CA generation that signed the certs it is currently presenting (i.e. the CA generation currently in use).
+
+When using cert-manager for issuing certificates Strimzi will continue to use these generations in the same way.
+Rather than directly mounting cert-manager Secrets and the user-provided CA cert Secret, Strimzi will copy the certificates into its own existing Secrets.
+
+#### Handling Cluster CA trust rollout
+
+When a new Kafka cluster is created Strimzi will copy the CA public cert from the Secret identified in `clusterCa.publicCert` and `clientsCa.publicCert` to the `<CLUSTER_NAME>-cluster-ca-cert` and `<CLUSTER_NAME>-clients-ca-cert` Secrets respectively.
+Strimzi will add three annotations to the `<CLUSTER_NAME>-cluster-ca-cert` Secret:
+* `strimzi.io/ca-cert-hash` with the value being a hash of the certificate.
+* `strimzi.io/cluster-ca-key-generation` initially set to 0.
+* `strimzi.io/cluster-ca-cert-generation` initially set to 0.
+
+During a reconciliation Strimzi will check the hash of the certificate stored in the user's CA public cert Secret to see if an update is needed.
+If the certificate has changed Strimzi will rename the existing cert in the `<CLUSTER_NAME>-cluster-ca-cert` Secret to `ca-YYYY-MM-DDTHH-MM-SSZ.crt` and copy over the new certificate.
+It will also update the `strimzi.io/ca-cert-hash` and increment the `strimzi.io/cluster-ca-key-generation` and `strimzi.io/cluster-ca-cert-generation` annotations.
```

Comment:
Will the hash be based on the source Secret value or on the certificate itself? I.e. would for example some newline change trigger this? Or will it be safe against it?

Reply:
We would use the same mechanism as we have today, which is to extract the cert as an x509 certificate, then calculate the hash, so it will be safe against any extra whitespace in the source Secret value.